### PR TITLE
Parallel download

### PR DESCRIPTION
**Pull request by agalitsyna**
I've parallelized sra download/parsing and revived some forgotten, but useful features (e.g. silent download). I also designed tests for parallelization and run the old ones to prove that I have not introduced any errors.

Hope you find this useful. I'll be happy to accept your suggestions on how to improve this part.

### CODE REVIEW

**Comment by guma44** [comment]
Hi, thanks a lot for this pull request. I really like the tests you added. I am in the process of reviewing and testing the code. The tests on Travis are not passing because there is no Bio module in requirements.txt file. It should be added - in the past, it was optional because most people would not use it to download SRA files anyway. This time it has to be included because tests are based on it. I also see some minor style issues but this is to be fixed in the end.

**Comment by antonkulaga** [comment]
@agalitsyna looks awesome, looking forward to seeing it merged!

**Comment by agalitsyna** [comment]
Hi, thanks a lot for the review! 
Can I fix the Bio module requirement and minor style issues and do pull request once more? 
If so, I'd be grateful for a general description of style issues.

**Comment by guma44** [comment]
Of course, you can. Concerning the style, whenever I can I try to follow pep8 so when you use any linter (sublime anaconda, atom python-ide, autopep8 or pycodestyle) should give you some hints. In the code, I live some marks too.

File: GEOparse/GEOTypes.py [hunk 1]
```diff
@@ -609,50 +616,36 @@ def download_SRA(self, email, directory='./', **kwargs):
                     # the directory name cannot contain many of the signs
                     re.sub(name_regex, '_', self.metadata['title'][0]))))
 
-            utils.mkdir_p(os.path.abspath(directory_path))
+        if nproc==1 and not return_list:
```
Comment (guma44) [inline GEOparse/GEOTypes.py]: In pep8 there are spaces around operators.

File: GEOparse/GEOTypes.py [hunk 2]
```diff
@@ -609,50 +616,36 @@ def download_SRA(self, email, directory='./', **kwargs):
...
-                    downloaded_paths += glob.glob(os.path.join(
-                        directory_path,
-                        "%s*.%s.gz" % (sra_run, filetype)
-                    ))
-                    if not keep_sra and filetype != 'sra':
-                        # Delete sra file
-                        os.unlink(filepath)
-                    else:
-                        downloaded_paths.append(filepath)
+                filepath = utils.download_unpack_SRA(path, ftpaddres,
```
Comment (guma44) [inline GEOparse/GEOTypes.py]: I would try to keep all function parameters in one line or all in separate lines. In some case I see I am not following this but this is also my mistake.

**Comment by agalitsyna** [comment]
Basically, I've completed code style review (not only for parallelization part) and added biopython to the requirements. I've also fixed some minor bugs that I've encountered. I've run a setup from scratch of GEOparser on a blank linux system, both installation and tests are ok. 

I avoided following some PEP8 recommendations that were already ignored in GEOparse (e.g. no capital letters in function names).

I'm confused by two things: 
1. Mixed usage of exceptions (logger.error vs GEOTypes defined exception classes). Is there any substantial logic for it? 
For example, I had to remove NoSRAToolkitException, because the SRA unpacking step was moved to utils.py, where there are no exception classes defined. Hope it's not a crime.

2. Usage of e-mail is required by my tests with SRA download, so I put my unused e-mail for that purpose in the code. Is it convenient?

File: GEOparse/GEOparse.py [hunk 3]
```diff
@@ -305,13 +305,7 @@ def parse_GDS_columns(lines, subsets):
...
     for subsetname, subset in iteritems(subsets):
         for expid in subset.metadata["sample_id"][0].split(","):
-            try:
-                subset_type = subset.get_type()
-                subset_ids[subset_type][expid] = \
-                    subset.metadata['description'][0]
-            except Exception as err:
-                logger.error("Error processing subsets: %s for subset %s" % (
-                    subset.get_type(), subsetname))
+            pass
```
[the remark this answers is not in the document]
Comment (agalitsyna) [reply GEOparse/GEOparse.py]: This pass is suspended in the limbo and obscurity, I presume. My bad, fixed now.

[the remark this answers is not in the document]
Comment (agalitsyna) [reply GEOparse/GEOparse.py]: Should I create pull request one more time?

[the remark this answers is not in the document]
Comment (guma44) [reply GEOparse/GEOparse.py]: I think I could fix it since I have merged it already.

[the remark this answers is not in the document]
Comment (agalitsyna) [reply GEOparse/GEOparse.py]: Thanks. Please, take look at one more thing I was too radical about https://github.com/agalitsyna/GEOparse/commit/720d40ea373cfef9bfb41d0b6d796c3d3aabf336

[the remark this answers is not in the document]
Comment (guma44) [reply GEOparse/GEOparse.py]: Indeed, I was looking into this pass and I noticed that. I will roll this back.

**Pull request merged**